cli/tools/registry/mod.rs

```diff
@@ -859,6 +910,45 @@ pub async fn publish(
     prepared_data.publish_order_graph,
     prepared_data.package_by_name,
     auth_method,
+    publish_flags.provenance,
   )
-  .await
+  .await?;
+
+  Ok(())
+}
+
+#[derive(Deserialize)]
+struct ManifestEntry {
+  checksum: String,
+}
+
+#[derive(Deserialize)]
+struct VersionManifest {
+  manifest: HashMap<String, ManifestEntry>,
+}
+
+fn verify_version_manifest(
+  meta_bytes: &[u8],
+  package: &PreparedPublishPackage,
+) -> Result<(), AnyError> {
+  let manifest = serde_json::from_slice::<VersionManifest>(meta_bytes)?;
+  for (path, entry) in manifest.manifest {
+    // Verify each path with the files in the tarball.
+    let file = package
+      .tarball
+      .files
+      .iter()
+      .find(|f| f.path_str == path.as_str());
+    if let Some(file) = file {
+      if file.hash != entry.checksum {
+        bail!(
+          "Checksum mismatch for {}: expected {}, got {}",
+          path,
+          entry.checksum,
+          file.hash
+        );
+      }
+    }
+  }
+  Ok(())
```

Comment:
Maybe this should error for any extra or missing items found in `VersionManifest`?

Reply:
I think it's fine. full struct verification is the job of deno_graph - this bit here is to make sure the registry is not tamperying with the uploaded package files.

Reply:
What if the registry has changed an export to a new file? It looks like this code would say that's fine unless I'm misreading it. Why would the registry be adding or removing files?

Reply:
Yeah, agreed - this code should:
- check that there are no additional files
- check that the `exports` line up

Reply:
I am not concerned about the `moduleGraph1` being tampered with because it alone does not result in code execution, it is only a preloading optimization.

Reply:
Added those checks.